src/ophyd_async/core/device.py

```diff
@@ -39,6 +41,13 @@ def name(self) -> str:
         """Return the name of the Device"""
         return self._name
 
+    @property
+    @lru_cache(1)
+    def log(self):
+        return LoggerAdapter(
+            getLogger("ophyd_async.devices"), {"ophyd_async_device_name": self.name}
```

Comment:
Is this guaranteed to be run for the first time after the name has been finalized? Naively I'd assume it's possible for some code to access this prior to name finalization (without any way to check here) but I'm not familiar enough with the initialization flow to know that for sure.

If this is guaranteed to not be run prior to name finalization, then this PR looks good to me.

Reply:
Thanks, good point. If we clear the cache during `set_name`, it should guarantee that we have the latest version for `name` in the log